### PR TITLE
Updates and cleanup in self hosted compiler

### PR DESCRIPTION
**Pull request by Vexu**
### Changes
- use enum literals
- remove `await (async .. catch unreachable)` pattern
- cleanup `target.zig` into `util.zig` and use `std.Target`
- update use of `std.event`

### TODO after this
* [x] update `std.event.fs`
* [ ] review and update use of async functions

### CODE REVIEW

**Comment by andrewrk** [comment]
Does this get it building again successfully? If so we could re-enable building it as part of the test suite:

https://github.com/ziglang/zig/blob/29d7b5a80c9faf640c3db0de14cd229e90b2d8c3/build.zig#L75-L78

**Comment by Vexu** [comment]
> Does this get it building again successfully? If so we could re-enable building it as part of the test suite:
> 
> https://github.com/ziglang/zig/blob/29d7b5a80c9faf640c3db0de14cd229e90b2d8c3/build.zig#L75-L78

It should.

**Comment by Vexu** [comment]
> It should.

Apparently only on Linux. Can someone on Mac or Windows give detailed error logs?

**Comment by andrewrk** [comment]
> Apparently only on Linux. Can someone on Mac or Windows give detailed error logs?

I can help you with that in a bit. If you click on the "Details" link in the CI above, do you see the error? It should be open access with no login required.

**Comment by andrewrk** [comment]
Oh, I can tell you what's going on here though. Evented I/O mode integration isn't hooked up with non-Linux operating systems yet. Specifically, the `waitUntilFdIsReadable` and related functions don't have non-Linux implementations yet.

It would still be helpful to enable building the self-hosted compiler as part of the test suite when the target is Linux though. That's still more test coverage than before, as we work towards full coverage and full support.

File: build.zig [hunk 1]
```diff
@@ -53,7 +53,10 @@ pub fn build(b: *Builder) !void {
     ctx.llvm = try findLLVM(b, ctx.llvm_config_exe);
 
     var test_stage2 = b.addTest("src-self-hosted/test.zig");
-    test_stage2.setBuildMode(builtin.Mode.Debug);
+    if (builtin.os == .linux) {
```
Comment (andrewrk) [inline build.zig]: Did you meant to put this over at line 80?

**Comment by andrewrk** [comment]
Oh, nice you even got the stage2 tests passing again. I'm guessing for the Windows failure, this code will need to chop off the extra `.lib` in the library names that we get from `llvm-config`:

https://github.com/ziglang/zig/blob/29d7b5a80c9faf640c3db0de14cd229e90b2d8c3/build.zig#L229-L242

I'd be happy to help with this; I have a Windows machine sitting right over here.

File: build.zig [hunk 2]
```diff
@@ -98,11 +99,7 @@ pub fn build(b: *Builder) !void {
 
     const test_stage2_step = b.step("test-stage2", "Run the stage2 compiler tests");
     test_stage2_step.dependOn(&test_stage2.step);
-
-    // TODO see https://github.com/ziglang/zig/issues/1364
-    if (false) {
-        test_step.dependOn(test_stage2_step);
-    }
+    test_step.dependOn(test_stage2_step);
```
Comment (andrewrk) [inline build.zig]: I don't think the stage2 tests are ready to be resurrected yet, if you look at `src-self-hosted/test.zig`, it creates its own event loop, and tries to do evented I/O, but `zig test` doesn't support that yet, so it's this weird mixture of evented and blocking (which amazingly apparently did work successfully on Linux and macOS)

**Pull request merged**